internal/app/secretless/tcp_proxy_service.go

```diff
@@ -0,0 +1,147 @@
+package secretless
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"reflect"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+
+	"github.com/cyberark/secretless-broker/internal"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/tcp"
+)
+
+// Zeroizes the values of the fetched credentials. We don't want to
+// rely on garbage collection for this (it might be slow and/or only free them) so
+// we manually clear
+func zeroizeCredentials(backendCredentials map[string][]byte) {
+	for _, rawCredential := range reflect.ValueOf(backendCredentials).MapKeys() {
+		credential := rawCredential.String()
+		for byteIndex := range backendCredentials[credential] {
+			backendCredentials[credential][byteIndex] = 0
+		}
+	}
+}
+
+func duplexStream(source io.ReadWriter, destination io.ReadWriter) <- chan error {
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- stream(source, destination)
+	}()
+	go func() {
+		errChan <- stream(destination, source)
+	}()
+	return errChan
+}
+
+func stream(source io.Reader, destination io.Writer) error {
+	for {
+		if _, err := io.Copy(destination, source); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+type tcpProxyService struct {
+	connector          tcp.Connector
+	resolveCredentials credentialsResolver
+	// TODO: use logger to report errors
+	// logger 			   log.Logger
+	listener           net.Listener
+	done               bool
+	error              chan error
+}
+
+// NewTCPProxyService constructs a new instance of a TCP ProxyService. The
+// constructor takes a TCP Connector, CredentialResolver and Listener.
+// A TCP ProxyService is able to Connect with Credentials then subsequently stream
+// bytes between client and target service
+func NewTCPProxyService(
+	connector tcp.Connector,
+	resolveCredentials credentialsResolver,
+	listener net.Listener,
+	) (internal.ProxyService, error) {
+	errors := validation.Errors{}
+
+	if connector == nil {
+		errors["connector"] = fmt.Errorf("connector cannot be nil")
+	}
+	if resolveCredentials == nil {
+		errors["resolveCredentials"] = fmt.Errorf("resolveCredentials cannot be nil")
+	}
+	if listener == nil {
+		errors["listener"] = fmt.Errorf("listener cannot be nil")
+	}
+
+	if err := errors.Filter(); err != nil {
+		return nil, err
+	}
+
+	return &tcpProxyService{
+		connector:          connector,
+		resolveCredentials: resolveCredentials,
+		listener:           listener,
+		done:               false,
+		error:              make(chan error),
+	}, nil
+}
+
+type credentialsResolver func() (map[string][]byte, error)
+
+func (ps *tcpProxyService) handleConnection(clientConn net.Conn) error {
```

Comment:
Not: Not a big deal and I know not everyone agrees with these but `ps` is really hard to identify as a tcpProxyService type. Any way to just change `ps`->`proxy`?